Mathlib/GroupTheory/GroupAction/FixedPoints.lean

```diff
@@ -0,0 +1,286 @@
+def movedBy (m : M) : Set α := { a : α | m • a ≠ a }
+
+variable {α}
+
+@[to_additive (attr := simp)]
+theorem mem_movedBy {m : M} {a : α} : a ∈ movedBy α m ↔ m • a ≠ a :=
+  Iff.rfl
+
+@[to_additive]
+theorem movedBy_eq_compl_fixedBy {m : M} : movedBy α m = (fixedBy α m)ᶜ := rfl
+
+@[to_additive]
+theorem fixedBy_eq_compl_movedBy {m : M} : fixedBy α m = (movedBy α m)ᶜ := by
+  rw [← compl_compl (fixedBy α m), movedBy_eq_compl_fixedBy]
+
+@[to_additive]
+theorem not_mem_fixedBy_iff_mem_movedBy {m : M} {a : α} : a ∉ fixedBy α m ↔ a ∈ movedBy α m :=
+  Iff.rfl
+
+@[to_additive]
+theorem not_mem_movedBy_iff_mem_fixedBy {m : M} {a : α} : a ∉ movedBy α m ↔ a ∈ fixedBy α m := by
+  rw [movedBy_eq_compl_fixedBy, Set.not_mem_compl_iff]
+
+variable (α)
+
+/-- In a multiplicative group action, the points moved by `g` are also moved by `g⁻¹` -/
+@[to_additive "In an additive group action, the points moved by `g` are also moved by `g⁻¹`"]
+theorem movedBy_eq_movedBy_inv (g : G) : movedBy α g = movedBy α g⁻¹ := by
+  repeat rw [movedBy_eq_compl_fixedBy]
+  rw [fixedBy_eq_fixedBy_inv]
+
+variable {α}
+
+@[to_additive]
+theorem smul_mem_movedBy_iff_mem_movedBy {a : α} {g : G} :
+    g • a ∈ movedBy α g ↔ a ∈ movedBy α g := by
+  repeat rw [← not_mem_fixedBy_iff_mem_movedBy]
+  rw [smul_mem_fixedBy_iff_mem_fixedBy]
+
+@[to_additive]
+theorem smul_inv_mem_movedBy_iff_mem_movedBy {a : α} {g : G} :
+    g⁻¹ • a ∈ movedBy α g ↔ a ∈ movedBy α g :=
+  (movedBy_eq_movedBy_inv α g) ▸ smul_mem_movedBy_iff_mem_movedBy
+
+@[to_additive]
+theorem movedBy_pow_subset_movedBy (g : G) (j : ℤ) :
+    movedBy α (g^j) ⊆ movedBy α g := by
+  repeat rw [movedBy_eq_compl_fixedBy]
+  rw [Set.compl_subset_compl]
+  exact fixedBy_subset_fixedBy_pow g j
+
+variable (M α)
+
+@[to_additive (attr := simp)]
+theorem movedBy_one_eq_empty : movedBy α (1 : M) = ∅ := by
+  rw [movedBy_eq_compl_fixedBy, fixedBy_one_eq_univ, Set.compl_univ]
+
+variable {M}
+
+@[to_additive]
+theorem movedBy_mul (m₁ m₂ : M) : movedBy α (m₁ * m₂) ⊆ movedBy α m₁ ∪ movedBy α m₂ := by
+  repeat rw [movedBy_eq_compl_fixedBy]
+  rw [← Set.compl_inter, Set.compl_subset_compl]
+  exact fixedBy_mul α m₁ m₂
+
+@[to_additive]
+theorem movedBy_conj (g h : G) :
+    movedBy α (h * g * h⁻¹) = (fun a => h⁻¹ • a) ⁻¹' movedBy α g := by
+  repeat rw [movedBy_eq_compl_fixedBy]
+  rw [Set.preimage_compl, fixedBy_conj]
+
+@[to_additive]
+theorem movedBy_conj' {g h : G} :
+    h • movedBy α g = movedBy α (h * g * h⁻¹) := by
+  rw [movedBy_conj, Set.preimage_smul, inv_inv]
+
+end MovedBy
+
+section Image
+
+/-- If `s` is a superset of `movedBy α g`, then `g` cannot move a point outside of `s`. -/
+@[to_additive "If `s` is a superset of `movedBy α g`, then `g` cannot move a point outside of `s`."]
+theorem smul_pow_mem_of_movedBy_subset {a : α} {s : Set α} {g : G} (s_subset : movedBy α g ⊆ s)
+    (j : ℤ) : g^j • a ∈ s ↔ a ∈ s := by
+  rcases (Classical.em (a ∈ movedBy α (g^j))) with a_moved | a_fixed
```

Comment:
The tactic `by_cases` does exactly this and it is nicer.

Reply:
I do miss the convenience of NNG's `cases`, since I never remember the name of the generated subgoals, and you need to name them explicitely to be able to access the generated hypothesis

Reply:
I mean you do this
```lean
/-- If `movedBy α g ⊆ s`, then `g` cannot move a point of `s` outside of `s`. -/
@[to_additive "If `movedBy α g ⊆ s`, then `g` cannot move a point of `s` outside of `s`."]
theorem smul_zpow_mem_of_movedBy_subset {a : α} {s : Set α} {g : G} (s_subset : movedBy α g ⊆ s)
    (j : ℤ) : g ^ j • a ∈ s ↔ a ∈ s := by
  by_cases h :  a ∈ movedBy α (g ^ j)
  · constructor
    all_goals (intro; apply s_subset)
    all_goals apply movedBy_zpow_subset_movedBy g j
    · exact h
    · rw [smul_mem_movedBy_iff_mem_movedBy]
      exact h
  · rw [not_mem_movedBy_iff_mem_fixedBy, mem_fixedBy] at h
    rw [h]
```